File: ios/BitBoxBleTransport.swift
import Bitboxnative
import CoreBluetooth
import Foundation

private let bitboxBleServiceUUID = CBUUID(string: "e1511a45-f3db-44c0-82b8-6c880790d1f1")
private let bitboxBleWriterUUID = CBUUID(string: "799d485c-d354-4ed0-b577-f8ee79ec275a")
private let bitboxBleReaderUUID = CBUUID(string: "419572a5-9f53-4eb1-8db7-61bcab928867")
private let bitboxBleProductUUID = CBUUID(string: "9d1c9a77-8b03-4e49-8053-3955cda7da93")
private let bitboxBleMaxCharacteristicLength = 5 * 64

struct BitBoxNativeError: LocalizedError {
  let message: String

  var errorDescription: String? {
    message
  }
}

struct BitBoxBleProductInfo: Decodable {
  let product: String
  let version: String

  enum CodingKeys: String, CodingKey {
    case product = "p"
    case version = "v"
  }
}

final class BitBoxBleTransport: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate,
  BitcoinerlabBitBoxBitboxnativeMobileTransportProtocol
{
  private let queue = DispatchQueue(label: "com.bitcoinerlab.bitboxreactnative.ble")
  private let poweredOnSemaphore = DispatchSemaphore(value: 0)
  private let readySemaphore = DispatchSemaphore(value: 0)
  private let readSemaphore = DispatchSemaphore(value: 0)
  private let readBufferLock = NSLock()
  private let operationTimeoutMs: Int

  private var centralManager: CBCentralManager?
  private var targetDeviceId: String?
  private var peripheral: CBPeripheral?
  private var writerCharacteristic: CBCharacteristic?
  private var readerCharacteristic: CBCharacteristic?
  private var productCharacteristic: CBCharacteristic?
  private var readerReady = false
  private var readySignaled = false
  private var closed = false
  private var connectionError: Error?
  private var readBuffer = Data()
  private var maxWriteLength = 64

  private(set) var productInfo: BitBoxBleProductInfo?

  init(timeoutMs: Int) {
    operationTimeoutMs = max(timeoutMs, 1)
    super.init()
  }

  func connect(deviceId: String?) throws -> BitBoxBleProductInfo {
    let deadline = DispatchTime.now() + .milliseconds(operationTimeoutMs)
    queue.sync {
      targetDeviceId = deviceId?.uppercased()
      centralManager = CBCentralManager(delegate: self, queue: queue)
    }

    if poweredOnSemaphore.wait(timeout: deadline) == .timedOut {
      try? close()
      throw BitBoxNativeError(message: "Timed out waiting for Bluetooth to become available")
    }
    if let error = currentConnectionError() {
      throw error
    }

    queue.sync {
      centralManager?.scanForPeripherals(withServices: [bitboxBleServiceUUID], options: nil)
    }

    if readySemaphore.wait(timeout: deadline) == .timedOut {
      try? close()
      throw BitBoxNativeError(message: "Timed out waiting for a BitBox Nova BLE device")
    }
    if let error = currentConnectionError() {
      throw error
    }
    guard let productInfo = queue.sync(execute: { self.productInfo }) else {
      throw BitBoxNativeError(message: "BitBox Nova BLE product information was not received")
    }
    return productInfo
  }

  func centralManagerDidUpdateState(_ central: CBCentralManager) {
    switch central.state {
    case .poweredOn:
      poweredOnSemaphore.signal()
    case .unauthorized:
      fail("Bluetooth permission was denied")
    case .poweredOff:
      fail("Bluetooth is powered off")
    case .unsupported:
      fail("Bluetooth is not supported on this device")
    case .resetting, .unknown:
      break
    @unknown default:
      fail("Bluetooth entered an unknown state")
    }
  }

  func centralManager(
    _ central: CBCentralManager,
    didDiscover peripheral: CBPeripheral,
    advertisementData: [String: Any],
    rssi RSSI: NSNumber
  ) {
    _ = (advertisementData, RSSI)
    if let targetDeviceId = targetDeviceId,
      peripheral.identifier.uuidString.uppercased() != targetDeviceId
    {
      return
    }
    self.peripheral = peripheral
    peripheral.delegate = self
    central.stopScan()
    central.connect(peripheral, options: nil)
  }

  func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
    peripheral.discoverServices([bitboxBleServiceUUID])
  }

  func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
    _ = central
    fail(error?.localizedDescription ?? "Failed to connect to BitBox Nova BLE peripheral")
  }

  func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
    _ = (central, peripheral)
    if !closed {
      fail(error?.localizedDescription ?? "BitBox Nova BLE peripheral disconnected")
    }
    readSemaphore.signal()
  }

  func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
    if let error = error {
      fail(error.localizedDescription)
      return
    }
    guard let service = peripheral.services?.first(where: { $0.uuid == bitboxBleServiceUUID }) else {
      fail("BitBox Nova BLE service was not found")
      return
    }
    peripheral.discoverCharacteristics(
      [bitboxBleWriterUUID, bitboxBleReaderUUID, bitboxBleProductUUID],
      for: service
    )
  }

  func peripheral(
    _ peripheral: CBPeripheral,
    didDiscoverCharacteristicsFor service: CBService,
    error: Error?
  ) {
    _ = service
    if let error = error {
      fail(error.localizedDescription)
      return
    }
    for characteristic in service.characteristics ?? [] {
      switch characteristic.uuid {
      case bitboxBleWriterUUID:
        writerCharacteristic = characteristic
        let mtuLength = peripheral.maximumWriteValueLength(for: .withoutResponse)
        maxWriteLength = max(64, (mtuLength / 64) * 64)
      case bitboxBleReaderUUID:
        readerCharacteristic = characteristic
        peripheral.setNotifyValue(true, for: characteristic)
      case bitboxBleProductUUID:
        productCharacteristic = characteristic
        peripheral.setNotifyValue(true, for: characteristic)
      default:
        break
      }
    }
    if writerCharacteristic == nil || readerCharacteristic == nil || productCharacteristic == nil {
      fail("BitBox Nova BLE characteristics were not found")
    }
  }

  func peripheral(_ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?) {
    _ = peripheral
    if let error = error {
      fail(error.localizedDescription)
      return
    }
    if characteristic.uuid == bitboxBleReaderUUID {
      readerReady = characteristic.isNotifying
      checkReady()
    }
  }

  func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
    _ = peripheral
    if let error = error {
      fail(error.localizedDescription)
      return
    }
    guard let value = characteristic.value else {
      return
    }

    if characteristic.uuid == bitboxBleReaderUUID {
      readBufferLock.lock()
      readBuffer.append(value)
      readBufferLock.unlock()
      readSemaphore.signal()
      return
    }

    if characteristic.uuid == bitboxBleProductUUID {
      if value.isEmpty {
        return
      }
      do {
        productInfo = try JSONDecoder().decode(BitBoxBleProductInfo.self, from: value)
        checkReady()
      } catch {
        fail("Failed to parse BitBox Nova product information")
      }
    }
  }

  func read(_ n: Int) throws -> Data {
    if n <= 0 {
      return Data()
    }

    let deadline = DispatchTime.now() + .milliseconds(operationTimeoutMs)
    var data = Data()
    while data.count < n {
      readBufferLock.lock()
      if !readBuffer.isEmpty {
        let count = min(n - data.count, readBuffer.count)
        data.append(readBuffer.prefix(count))
        readBuffer.removeSubrange(0..<count)
        readBufferLock.unlock()
        continue
      }
      readBufferLock.unlock()

      if let error = currentConnectionError() {
        throw error
      }
      if !isConnected() {
        throw BitBoxNativeError(message: "BitBox Nova BLE peripheral is not connected")
      }
      if readSemaphore.wait(timeout: deadline) == .timedOut {
        throw BitBoxNativeError(message: "Timed out reading from BitBox Nova BLE peripheral")
      }
    }
    return data
  }

  func write(_ data: Data?, ret0_: UnsafeMutablePointer<Int>?) throws {
    guard let data = data, !data.isEmpty else {
      ret0_?.pointee = 0
      return
    }

    let written = try queue.sync { () throws -> Int in
      if let connectionError = connectionError {
        throw connectionError
      }
      guard !closed,
        let peripheral = peripheral,
        let writerCharacteristic = writerCharacteristic
      else {
        throw BitBoxNativeError(message: "BitBox Nova BLE peripheral is not connected")
      }
      let count = min(bitboxBleMaxCharacteristicLength, maxWriteLength, data.count)
      guard count > 0 else {
        return 0
      }
      peripheral.writeValue(Data(data.prefix(count)), for: writerCharacteristic, type: .withResponse)
      return count
    }
    ret0_?.pointee = written
  }

  func close() throws {
    queue.sync {
      closed = true
      centralManager?.stopScan()
      if let peripheral = peripheral {
        centralManager?.cancelPeripheralConnection(peripheral)
      }
      peripheral = nil
      writerCharacteristic = nil
      readerCharacteristic = nil
      productCharacteristic = nil
      readerReady = false
    }
    readSemaphore.signal()
  }

  private func currentConnectionError() -> Error? {
    queue.sync {
      connectionError
    }
  }

  private func isConnected() -> Bool {
    queue.sync {
      !closed && peripheral != nil && writerCharacteristic != nil && readerReady
    }
  }

  private func fail(_ message: String) {
    fail(BitBoxNativeError(message: message))
  }

  private func fail(_ error: Error) {
    if connectionError == nil {
      connectionError = error
    }
    poweredOnSemaphore.signal()
    readySemaphore.signal()
    readSemaphore.signal()
  }

  private func checkReady() {
    guard !readySignaled, writerCharacteristic != nil, readerReady, productInfo != nil else {
      return
    }
    readySignaled = true
    readySemaphore.signal()
  }
}

File: android/src/main/java/com/bitcoinerlab/bitboxreactnative/BitBoxBleTransport.kt
package com.bitcoinerlab.bitboxreactnative

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Build
import android.os.ParcelUuid
import com.bitcoinerlab.bitboxreactnative.go.bitboxnative.MobileTransport
import java.io.ByteArrayOutputStream
import java.util.UUID
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import kotlin.math.max
import kotlin.math.min
import org.json.JSONObject

private val BITBOX_BLE_SERVICE_UUID: UUID = UUID.fromString("e1511a45-f3db-44c0-82b8-6c880790d1f1")
private val BITBOX_BLE_WRITER_UUID: UUID = UUID.fromString("799d485c-d354-4ed0-b577-f8ee79ec275a")
private val BITBOX_BLE_READER_UUID: UUID = UUID.fromString("419572a5-9f53-4eb1-8db7-61bcab928867")
private val BITBOX_BLE_PRODUCT_UUID: UUID = UUID.fromString("9d1c9a77-8b03-4e49-8053-3955cda7da93")
private val CLIENT_CHARACTERISTIC_CONFIG_UUID: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
private const val BITBOX_BLE_MAX_CHARACTERISTIC_LENGTH = 5 * 64

class BitBoxBleTransport(
  private val context: Context,
  private val timeoutMs: Int,
  private val deviceId: String?
) : MobileTransport {
  private val readLock = Object()
  private val writeLock = Object()
  private val readyLatch = CountDownLatch(1)
  private val normalizedDeviceId = deviceId?.uppercase()

  @Volatile private var scanner: android.bluetooth.le.BluetoothLeScanner? = null
  @Volatile private var gatt: BluetoothGatt? = null
  @Volatile private var writerCharacteristic: BluetoothGattCharacteristic? = null
  @Volatile private var readerReady = false
  @Volatile private var closed = false
  @Volatile private var connectionError: Throwable? = null
  @Volatile private var productInfo: BitBoxProductInfo? = null
  @Volatile private var maxWriteLength = 20

  private var readBuffer = ByteArray(0)
  private var pendingWriteLatch: CountDownLatch? = null
  private var pendingWriteError: Throwable? = null

  @SuppressLint("MissingPermission")
  fun connect(): BitBoxProductInfo {
    val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
      ?: throw BitBoxNativeException("Bluetooth manager is not available")
    val adapter = bluetoothManager.adapter
      ?: throw BitBoxNativeException("Bluetooth is not supported on this device")
    if (!adapter.isEnabled) {
      throw BitBoxNativeException("Bluetooth is powered off")
    }
    scanner = adapter.bluetoothLeScanner
      ?: throw BitBoxNativeException("Bluetooth LE scanner is not available")

    val filter = ScanFilter.Builder()
      .setServiceUuid(ParcelUuid(BITBOX_BLE_SERVICE_UUID))
      .build()
    val settings = ScanSettings.Builder()
      .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
      .build()
    scanner?.startScan(listOf(filter), settings, scanCallback)

    if (!readyLatch.await(timeoutMs.toLong(), TimeUnit.MILLISECONDS)) {
      close()
      throw BitBoxNativeException("Timed out waiting for a BitBox Nova BLE device")
    }
    currentConnectionError()?.let { throw it }
    return productInfo
      ?: throw BitBoxNativeException("BitBox Nova BLE product information was not received")
  }

  private val scanCallback = object : ScanCallback() {
    @SuppressLint("MissingPermission")
    override fun onScanResult(callbackType: Int, result: ScanResult) {
      val device = result.device ?: return
      if (normalizedDeviceId != null && device.address.uppercase() != normalizedDeviceId) {
        return
      }
      try {
        scanner?.stopScan(this)
      } catch (_: Throwable) {}
      gatt = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
      } else {
        device.connectGatt(context, false, gattCallback)
      }
    }

    override fun onScanFailed(errorCode: Int) {
      fail("Bluetooth LE scan failed with code $errorCode")
    }
  }

  private val gattCallback = object : BluetoothGattCallback() {
    @SuppressLint("MissingPermission")
    override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
      if (status != BluetoothGatt.GATT_SUCCESS) {
        fail("BitBox Nova BLE connection failed with GATT status $status")
        return
      }
      when (newState) {
        BluetoothProfile.STATE_CONNECTED -> {
          if (!gatt.requestMtu(517)) {
            gatt.discoverServices()
          }
        }
        BluetoothProfile.STATE_DISCONNECTED -> {
          if (!closed) fail("BitBox Nova BLE peripheral disconnected")
        }
      }
    }

    @SuppressLint("MissingPermission")
    override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
      if (status == BluetoothGatt.GATT_SUCCESS) {
        val payloadLength = max(20, mtu - 3)
        maxWriteLength = if (payloadLength >= 64) {
          min(BITBOX_BLE_MAX_CHARACTERISTIC_LENGTH, (payloadLength / 64) * 64)
        } else {
          payloadLength
        }
      }
      gatt.discoverServices()
    }

    @SuppressLint("MissingPermission")
    override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
      if (status != BluetoothGatt.GATT_SUCCESS) {
        fail("BitBox Nova BLE service discovery failed with GATT status $status")
        return
      }
      val service = gatt.getService(BITBOX_BLE_SERVICE_UUID)
      if (service == null) {
        fail("BitBox Nova BLE service was not found")
        return
      }
      writerCharacteristic = service.getCharacteristic(BITBOX_BLE_WRITER_UUID)
      val readerCharacteristic = service.getCharacteristic(BITBOX_BLE_READER_UUID)
      val productCharacteristic = service.getCharacteristic(BITBOX_BLE_PRODUCT_UUID)
      if (writerCharacteristic == null || readerCharacteristic == null || productCharacteristic == null) {
        fail("BitBox Nova BLE characteristics were not found")
        return
      }
      writerCharacteristic?.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
      enableNotifications(gatt, readerCharacteristic, "reader")
    }

    @SuppressLint("MissingPermission")
    override fun onDescriptorWrite(
      gatt: BluetoothGatt,
      descriptor: BluetoothGattDescriptor,
      status: Int
    ) {
      if (status != BluetoothGatt.GATT_SUCCESS) {
        val name = characteristicName(descriptor.characteristic)
        fail(
          "BitBox Nova BLE $name notification setup failed with GATT status $status " +
            "and properties ${descriptor.characteristic.properties}"
        )
        return
      }
      val characteristic = descriptor.characteristic
      if (characteristic.uuid == BITBOX_BLE_READER_UUID) {
        readerReady = true
        val productCharacteristic = gatt.getService(BITBOX_BLE_SERVICE_UUID)
          ?.getCharacteristic(BITBOX_BLE_PRODUCT_UUID)
        if (productCharacteristic == null) {
          fail("BitBox Nova BLE product characteristic was not found")
          return
        }
        enableNotifications(gatt, productCharacteristic, "product")
        return
      }
      if (characteristic.uuid == BITBOX_BLE_PRODUCT_UUID) {
        gatt.readCharacteristic(characteristic)
      }
    }

    override fun onCharacteristicChanged(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic
    ) {
      @Suppress("DEPRECATION")
      handleCharacteristicValue(characteristic, characteristic.value ?: ByteArray(0))
    }

    override fun onCharacteristicChanged(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
      value: ByteArray
    ) {
      handleCharacteristicValue(characteristic, value)
    }

    override fun onCharacteristicRead(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
      status: Int
    ) {
      if (status != BluetoothGatt.GATT_SUCCESS) return
      @Suppress("DEPRECATION")
      handleCharacteristicValue(characteristic, characteristic.value ?: ByteArray(0))
    }

    override fun onCharacteristicRead(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
      value: ByteArray,
      status: Int
    ) {
      if (status == BluetoothGatt.GATT_SUCCESS) {
        handleCharacteristicValue(characteristic, value)
      }
    }

    override fun onCharacteristicWrite(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
      status: Int
    ) {
      synchronized(writeLock) {
        if (status != BluetoothGatt.GATT_SUCCESS) {
          pendingWriteError = BitBoxNativeException("BitBox Nova BLE write failed with GATT status $status")
        }
        pendingWriteLatch?.countDown()
      }
    }
  }

  @SuppressLint("MissingPermission")
  private fun enableNotifications(
    gatt: BluetoothGatt,
    characteristic: BluetoothGattCharacteristic,
    name: String
  ) {
    val cccdValue = when {
      characteristic.properties and BluetoothGattCharacteristic.PROPERTY_NOTIFY != 0 ->
        BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
      characteristic.properties and BluetoothGattCharacteristic.PROPERTY_INDICATE != 0 ->
        BluetoothGattDescriptor.ENABLE_INDICATION_VALUE
      else -> {
        fail(
          "BitBox Nova BLE $name characteristic does not support notifications or indications " +
            "(properties ${characteristic.properties})"
        )
        return
      }
    }
    if (!gatt.setCharacteristicNotification(characteristic, true)) {
      fail(
        "Failed to enable BitBox Nova BLE $name notifications " +
          "(properties ${characteristic.properties})"
      )
      return
    }
    val descriptor = characteristic.getDescriptor(CLIENT_CHARACTERISTIC_CONFIG_UUID)
    if (descriptor == null) {
      fail("BitBox Nova BLE $name notification descriptor was not found")
      return
    }
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      val status = gatt.writeDescriptor(descriptor, cccdValue)
      if (status != 0) {
        fail(
          "Failed to write BitBox Nova BLE $name notification descriptor " +
            "(write status $status, properties ${characteristic.properties})"
        )
      }
    } else {
      @Suppress("DEPRECATION")
      descriptor.value = cccdValue
      @Suppress("DEPRECATION")
      if (!gatt.writeDescriptor(descriptor)) {
        fail(
          "Failed to write BitBox Nova BLE $name notification descriptor " +
            "(properties ${characteristic.properties})"
        )
      }
    }
  }

  private fun characteristicName(characteristic: BluetoothGattCharacteristic): String =
    when (characteristic.uuid) {
      BITBOX_BLE_READER_UUID -> "reader"
      BITBOX_BLE_PRODUCT_UUID -> "product"
      BITBOX_BLE_WRITER_UUID -> "writer"
      else -> characteristic.uuid.toString()
    }

  private fun handleCharacteristicValue(
    characteristic: BluetoothGattCharacteristic,
    value: ByteArray
  ) {
    when (characteristic.uuid) {
      BITBOX_BLE_READER_UUID -> {
        synchronized(readLock) {
          readBuffer += value
          readLock.notifyAll()
        }
      }
      BITBOX_BLE_PRODUCT_UUID -> {
        if (value.isEmpty()) return
        try {
          val json = JSONObject(String(value, Charsets.UTF_8))
          productInfo = BitBoxProductInfo(
            product = json.getString("p"),
            version = json.getString("v")
          )
          checkReady()
        } catch (error: Throwable) {
          fail(BitBoxNativeException("Failed to parse BitBox Nova product information", error))
        }
      }
    }
  }

  override fun read(n: Long): ByteArray {
    if (n <= 0) return ByteArray(0)
    if (n > Int.MAX_VALUE) throw BitBoxNativeException("BitBox BLE read length is too large")
    val output = ByteArrayOutputStream(n.toInt())
    val deadline = System.currentTimeMillis() + timeoutMs
    while (output.size() < n.toInt()) {
      currentConnectionError()?.let { throw it }
      synchronized(readLock) {
        if (readBuffer.isNotEmpty()) {
          val count = min(n.toInt() - output.size(), readBuffer.size)
          output.write(readBuffer, 0, count)
          readBuffer = readBuffer.copyOfRange(count, readBuffer.size)
          return@synchronized
        }
        val remainingMs = deadline - System.currentTimeMillis()
        if (remainingMs <= 0) {
          throw BitBoxNativeException("Timed out reading from BitBox Nova BLE peripheral")
        }
        readLock.wait(remainingMs)
      }
    }
    return output.toByteArray()
  }

  @SuppressLint("MissingPermission")
  override fun write(data: ByteArray): Long {
    if (data.isEmpty()) return 0
    currentConnectionError()?.let { throw it }
    val localGatt = gatt ?: throw BitBoxNativeException("BitBox Nova BLE peripheral is not connected")
    val characteristic = writerCharacteristic
      ?: throw BitBoxNativeException("BitBox Nova BLE writer characteristic is not available")
    val count = min(data.size, min(BITBOX_BLE_MAX_CHARACTERISTIC_LENGTH, maxWriteLength))
    if (count <= 0) return 0
    val chunk = data.copyOfRange(0, count)
    val latch = CountDownLatch(1)
    synchronized(writeLock) {
      if (pendingWriteLatch != null) {
        throw BitBoxNativeException("BitBox Nova BLE write already in progress")
      }
      pendingWriteError = null
      pendingWriteLatch = latch
    }
    val started = try {
      @Suppress("DEPRECATION")
      characteristic.value = chunk
      characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
      localGatt.writeCharacteristic(characteristic)
    } catch (error: Throwable) {
      synchronized(writeLock) {
        pendingWriteLatch = null
      }
      throw error
    }
    if (!started) {
      synchronized(writeLock) {
        pendingWriteLatch = null
      }
      throw BitBoxNativeException("Failed to write to BitBox Nova BLE peripheral")
    }
    if (!latch.await(timeoutMs.toLong(), TimeUnit.MILLISECONDS)) {
      synchronized(writeLock) {
        pendingWriteLatch = null
      }
      throw BitBoxNativeException("Timed out writing to BitBox Nova BLE peripheral")
    }
    synchronized(writeLock) {
      val error = pendingWriteError
      pendingWriteError = null
      pendingWriteLatch = null
      if (error != null) throw error
    }
    return count.toLong()
  }

  @SuppressLint("MissingPermission")
  override fun close() {
    closed = true
    try {
      scanner?.stopScan(scanCallback)
    } catch (_: Throwable) {}
    scanner = null
    try {
      gatt?.disconnect()
      gatt?.close()
    } catch (_: Throwable) {}
    gatt = null
    writerCharacteristic = null
    synchronized(readLock) {
      readLock.notifyAll()
    }
    synchronized(writeLock) {
      pendingWriteLatch?.countDown()
    }
    readyLatch.countDown()
  }

  private fun checkReady() {
    if (writerCharacteristic != null && readerReady && productInfo != null) {
      readyLatch.countDown()
    }
  }

  private fun currentConnectionError(): Throwable? = connectionError

  private fun fail(message: String) {
    fail(BitBoxNativeException(message))
  }

  private fun fail(error: Throwable) {
    if (connectionError == null) {
      connectionError = error
    }
    synchronized(readLock) {
      readLock.notifyAll()
    }
    synchronized(writeLock) {
      pendingWriteLatch?.countDown()
    }
    readyLatch.countDown()
  }
}
